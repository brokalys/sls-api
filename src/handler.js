import { ApolloServer, ApolloError } from 'apollo-server-lambda';
import {
  ApolloServerPluginInlineTrace,
  ApolloServerPluginInlineTraceDisabled,
  ApolloServerPluginUsageReporting,
  ApolloServerPluginUsageReportingDisabled,
} from 'apollo-server-core';
import Buildings from './data-sources/buildings';
import Properties from './data-sources/properties';
import loadUser from './lib/auth';
import AuhtDirective from './lib/auth-directive';
import Bugsnag from './lib/bugsnag';
import mysql from './lib/db';
import schema from './schema/schema.graphql';
import dbConfig from './db-config';
import resolvers from './resolvers';
import './knex-extensions';

const isDevMode = process.env.STAGE === 'dev';
const isTestMode = process.env.NODE_ENV === 'test';

export const server = new ApolloServer({
  typeDefs: schema,
  schemaDirectives: {
    auth: AuhtDirective,
  },
  resolvers,
  dataSources: () => ({
    buildings: new Buildings(dbConfig),
    properties: new Properties(dbConfig),
  }),
  tracing: isDevMode,
  playground: isDevMode,
  context: async ({ event, req, request, context }) => {
    const { requestContext } = event ||
      req || { requestContext: { identity: {} } };

    return {
      user: await loadUser(requestContext.identity.apiKeyId),
      invokedFunctionArn: context ? context.invokedFunctionArn : '',
    };
  },
  formatError: (error) => {
    if (
      (isDevMode && !isTestMode) ||
      error instanceof ApolloError ||
      error.originalError instanceof ApolloError ||
      error.originalError === undefined
    ) {
      return error;
    }

    if (!isTestMode) {
      console.log(error);
      Bugsnag.notify(error);
    }

    return new Error('An unexpected error occurred. Please try again later.');
  },
  plugins: [
    isDevMode
      ? ApolloServerPluginInlineTrace()
      : ApolloServerPluginInlineTraceDisabled(),
    !isDevMode && !isTestMode
      ? ApolloServerPluginUsageReporting({
          sendVariableValues: {
            exceptNames: ['email', 'unsubscribe_key'],
          },
          logger: {
            debug: () => {},
            info: () => {},
            warn: console.warn,
            error: console.error,
          },
          generateClientInfo: ({ context, request }) => {
            const headers = request.http && request.http.headers;
            const clientVersion =
              headers['apollographql-client-version'] || '1.0.0';

            if (!context.user) {
              return {
                clientName: 'Unauthorized',
                clientVersion,
              };
            }

            return {
              clientName: context.user.apiKey.name,
              clientVersion,
            };
          },
        })
      : ApolloServerPluginUsageReportingDisabled(),
    {
      requestDidStart(requestContext) {
        return {
          willSendResponse() {
            return mysql.end();
          },
        };
      },
    },
  ],
});

exports.graphqlHandler = server.createHandler({
  cors: {
    origin: '*',
  },
});
