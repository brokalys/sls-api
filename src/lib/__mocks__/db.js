export default {
  connect: jest.fn(),
  getClient: jest.fn(() => ({
    escape: jest.fn(),
  })),
  query: jest.fn(() => []),
  end: jest.fn(),
};
