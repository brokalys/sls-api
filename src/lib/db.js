import serverlessMysql from 'serverless-mysql';

const mysql = serverlessMysql({
  config: {
    host: process.env.DB_HOST,
    database: process.env.DB_DATABASE,
    user: process.env.DB_USERNAME,
    password: process.env.DB_PASSWORD,
  },
  maxRetries: 5,
});

if (process.env.NODE_ENV !== 'test') {
  mysql.connect();
}

export default mysql;
