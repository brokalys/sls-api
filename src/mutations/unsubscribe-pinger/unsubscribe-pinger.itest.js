import { createTestClient } from 'apollo-server-testing';

import { server } from 'handler';
import db from 'lib/db';

jest.mock('lib/db');

describe('Mutation: createProperty', () => {
  let mutate;

  beforeEach(() => {
    process.env.BROKALYS_PRIVATE_KEY = 'PRIVATE_KEY';

    const utils = createTestClient(server);
    mutate = utils.mutate;
  });

  test('unsubscribes an existing pinger', async () => {
    db.query.mockReturnValueOnce({
      affectedRows: 1,
    });

    const response = await mutate({
      mutation: `
        mutation {
          unsubscribePinger(
            id: 1
            unsubscribe_key: "test_123"
          )
        }
      `,
    });

    expect(response).toMatchSnapshot();
  });

  test('unsubscribes all pingers of an email', async () => {
    db.query.mockReturnValueOnce([
      {
        email: 'test@brokalys.com',
      },
    ]);
    db.query.mockReturnValueOnce({
      affectedRows: 1,
    });

    const response = await mutate({
      mutation: `
        mutation {
          unsubscribePinger(
            id: 1
            unsubscribe_key: "test_123"
            all: true
          )
        }
      `,
    });

    expect(response).toMatchSnapshot();
  });

  test('fails unsubscribing a pinger with wrong credentials, but still responds with status = true', async () => {
    db.query.mockReturnValueOnce({
      affectedRows: 0,
    });

    const response = await mutate({
      mutation: `
        mutation {
          unsubscribePinger(
            id: 1
            unsubscribe_key: "wrong_credentials"
          )
        }
      `,
    });

    expect(response).toMatchSnapshot();
  });

  test('fails unsubscribing with missing `id`', async () => {
    const response = await mutate({
      mutation: `
        mutation {
          unsubscribePinger(
            unsubscribe_key: "test_123"
          )
        }
      `,
    });

    expect(response).toMatchSnapshot();
  });

  test('fails unsubscribing with missing `unsubscribe_key`', async () => {
    const response = await mutate({
      mutation: `
        mutation {
          unsubscribePinger(
            id: 1
          )
        }
      `,
    });

    expect(response).toMatchSnapshot();
  });
});
