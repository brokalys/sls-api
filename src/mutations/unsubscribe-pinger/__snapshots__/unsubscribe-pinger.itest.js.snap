// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutation: createProperty fails unsubscribing a pinger with wrong credentials, but still responds with status = true 1`] = `
Object {
  "data": Object {
    "unsubscribePinger": true,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation: createProperty fails unsubscribing with missing \`id\` 1`] = `
Object {
  "data": undefined,
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "GRAPHQL_VALIDATION_FAILED",
      },
      "locations": Array [
        Object {
          "column": 11,
          "line": 3,
        },
      ],
      "message": "Field \\"unsubscribePinger\\" argument \\"id\\" of type \\"String!\\" is required, but it was not provided.",
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation: createProperty fails unsubscribing with missing \`unsubscribe_key\` 1`] = `
Object {
  "data": undefined,
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "GRAPHQL_VALIDATION_FAILED",
      },
      "locations": Array [
        Object {
          "column": 11,
          "line": 3,
        },
      ],
      "message": "Field \\"unsubscribePinger\\" argument \\"unsubscribe_key\\" of type \\"String!\\" is required, but it was not provided.",
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation: createProperty unsubscribes all pingers of an email 1`] = `
Object {
  "data": Object {
    "unsubscribePinger": true,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation: createProperty unsubscribes an existing pinger 1`] = `
Object {
  "data": Object {
    "unsubscribePinger": true,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation: createProperty unsubscribes an existing pinger with id-hash 1`] = `
Object {
  "data": Object {
    "unsubscribePinger": true,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
